ostd/src/arch/x86/iommu/dma_remapping/second_stage.rs

```diff
@@ -84,81 +83,101 @@ bitflags::bitflags! {
 
 #[derive(Debug, Clone, Copy, Pod, Default)]
 #[repr(C)]
-pub struct PageTableEntry(u64);
+pub struct PageTableEntry(usize);
 
-impl PageTableEntry {
-    const PHYS_MASK: u64 = 0xFFFF_FFFF_F000;
-    const PROP_MASK: u64 = !Self::PHYS_MASK & !PageTableFlags::LAST_PAGE.bits();
+/// Parse a bit-flag bits `val` in the representation of `from` to `to` in bits.
+macro_rules! parse_flags {
+    ($val:expr, $from:expr, $to:expr) => {
+        (($val as usize & $from.bits() as usize) >> $from.bits().ilog2() << $to.bits().ilog2())
+    };
 }
 
-impl PodOnce for PageTableEntry {}
-
-impl PageTableEntryTrait for PageTableEntry {
-    fn new_page(paddr: Paddr, level: PagingLevel, prop: PageProperty) -> Self {
-        let mut pte = Self(paddr as u64 & Self::PHYS_MASK | PageTableFlags::LAST_PAGE.bits());
-        pte.set_prop(prop);
-        pte
-    }
-
-    fn new_pt(paddr: Paddr) -> Self {
-        Self(
-            paddr as u64 & Self::PHYS_MASK
-                | PageTableFlags::READABLE.bits()
-                | PageTableFlags::WRITABLE.bits(),
-        )
-    }
+impl PageTableEntry {
+    const PHYS_MASK: usize = 0xffff_ffff_f000;
 
-    fn paddr(&self) -> Paddr {
-        (self.0 & Self::PHYS_MASK) as usize
+    fn is_present(&self) -> bool {
+        self.0 & (PteFlags::READABLE | PteFlags::WRITABLE).bits() != 0
```

Comment:
`PteFlags::LAST_PAGE | PteFlags::READABLE` because `new_page` will set `LAST_PAGE` and `new_pt` will set `READABLE`?

Otherwise, `is_present` will return false if a page is protected as non-readable and non-writable.